pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java

```diff
@@ -1651,118 +1651,123 @@ private void internalDeleteSubscriptionForNonPartitionedTopicForcefully(AsyncRes
+                                    asyncResponse.resume(new RestException(e));
+                                    return;
+                                }
+                            }
 
-                    FutureUtil.waitForAll(futures).handle((result, exception) -> {
-                        if (exception != null) {
-                            Throwable t = exception.getCause();
-                            if (t instanceof NotFoundException) {
-                                asyncResponse.resume(new RestException(Status.NOT_FOUND, "Subscription not found"));
-                                return null;
-                            } else {
-                                log.error("[{}] Failed to skip all messages {} {}",
-                                        clientAppId(), topicName, subName, t);
-                                asyncResponse.resume(new RestException(t));
+                            FutureUtil.waitForAll(futures).handle((result, exception) -> {
+                                if (exception != null) {
+                                    Throwable t = exception.getCause();
+                                    if (t instanceof NotFoundException) {
+                                        asyncResponse.resume(
+                                            new RestException(Status.NOT_FOUND, "Subscription not found"));
+                                        return null;
+                                    } else {
+                                        log.error("[{}] Failed to skip all messages {} {}",
+                                            clientAppId(), topicName, subName, t);
+                                        asyncResponse.resume(new RestException(t));
+                                        return null;
+                                    }
+                                }
+
+                                asyncResponse.resume(Response.noContent().build());
                                 return null;
-                            }
+                            });
+                        } else {
+                            internalSkipAllMessagesForNonPartitionedTopicAsync(asyncResponse, subName, authoritative);
                         }
-
-                        asyncResponse.resume(Response.noContent().build());
+                    }).exceptionally(ex -> {
+                        log.error("[{}] Failed to skip all messages for subscription {} on topic {}",
+                            clientAppId(), subName, topicName, ex);
+                        resumeAsyncResponseExceptionally(asyncResponse, ex);
                         return null;
                     });
-                } else {
-                    internalSkipAllMessagesForNonPartitionedTopic(asyncResponse, subName, authoritative);
                 }
             }).exceptionally(ex -> {
                 log.error("[{}] Failed to skip all messages for subscription {} on topic {}",
```

Comment:
We should use `ex.getCause()`

Reply:
OK, I see there is a better method `FutureUtil.unwrapCompletionException`